src/wechaty/plugin.py

```diff
@@ -0,0 +1,325 @@
+"""
+Python Wechaty - https://github.com/wechaty/python-wechaty
+
+Authors:    Huan LI (李卓桓) <https://github.com/huan>
+            Jingjing WU (吴京京) <https://github.com/wj-Mcat>
+
+2020-now @ Copyright Wechaty
+
+Licensed under the Apache License, Version 2.0 (the 'License');
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an 'AS IS' BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+from __future__ import annotations
+
+from abc import abstractmethod, ABCMeta
+from enum import Enum
+from typing import (
+    TYPE_CHECKING,
+    List,
+    Optional,
+    Dict,
+    Union
+)
+import re
+from datetime import datetime
+from copy import deepcopy
+from dataclasses import dataclass
+from collections import defaultdict, OrderedDict
+from wechaty_puppet import get_logger   # type: ignore
+
+if TYPE_CHECKING:
+    from wechaty_puppet import (
+        EventErrorPayload,
+        EventHeartbeatPayload,
+        EventReadyPayload,
+        ScanStatus
+    )
+    from wechaty import (
+        Room,
+        Friendship,
+        Contact,
+        Message,
+        Wechaty
+    )
+    from wechaty.user.room_invitation import RoomInvitation
+
+
+log = get_logger(__name__)
+
+
+@dataclass
+class WechatyPluginOptions:
```

Comment:
I'd like to suggest that we leave the `Options` schema defined by our plugin developers. (the TS has done this)

For example, my `EventLogger` Plugin will accept an array of strings, instead of a Map(dataclass)

Reply:
Never mind, I think the `dataclass` is good.

Reply:
`WechatyPluginOptions` is a base class but not abstract class, which contains common attributes for all plugins. If there is a plugin that requires custom attributes, So, they can create a `PluginOptions` class which inherit from `WechatyPluginOptions`.